### PR TITLE
ci: setup git tags

### PR DESCRIPTION
**Pull request by BioPhoton**
This PR includes the basic setup to version our packages.

A couple of things I want to point out here:
- I added all configurations of `version` to the  `targetDefaults`
  This reduces the duplicated options and leaves the `project.json` with just `"version": {}` as target
- I did NOT add `version` as a cache target under `tasksRunnerOptions`
  As the result is idempotent. It if there is no new change in the git history it will just pass.
- We could consider the history by adding tags, to restore a bit of the changes we did before versioning
  
  Todos:
  - [x] add `version` targets
  - [x] configure `nx.json`
  - [x] add github target

Action setup is done in another PR

related to #8

### CODE REVIEW

**Comment by BioPhoton** [comment]
Example of current readme:

The scopes and packages don't seem to be related, or we have a bad commit hygiene :D
I would like to discuss this asap.

---

# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

## 0.1.0 (2023-10-17)

### Features

- **cli:** initial collect command ([#45](https://github.com/flowup/quality-metrics-cli/issues/45)) ([ba048be](https://github.com/flowup/quality-metrics-cli/commit/ba048be5f3e9b4291ae6311051cda89403186795))
- **cli:** persist login and formatting options ([#47](https://github.com/flowup/quality-metrics-cli/issues/47)) ([6241fd7](https://github.com/flowup/quality-metrics-cli/commit/6241fd7412f33a8d40183b40f8ed7e39e85278dd))

## 0.1.0 (2023-10-17)

### Features

- **cli:** initial collect command ([#45](https://github.com/flowup/quality-metrics-cli/issues/45)) ([ba048be](https://github.com/flowup/quality-metrics-cli/commit/ba048be5f3e9b4291ae6311051cda89403186795))
- **cli:** persist login and formatting options ([#47](https://github.com/flowup/quality-metrics-cli/issues/47)) ([6241fd7](https://github.com/flowup/quality-metrics-cli/commit/6241fd7412f33a8d40183b40f8ed7e39e85278dd))
- **cli:** setup yargs for cli ([#42](https://github.com/flowup/quality-metrics-cli/issues/42)) ([37ea0a5](https://github.com/flowup/quality-metrics-cli/commit/37ea0a56c987ef38cf7e961d885c711b32de777a)), closes [#6](https://github.com/flowup/quality-metrics-cli/issues/6) [#38](https://github.com/flowup/quality-metrics-cli/issues/38)
- **cli:** use bundle-require instead of jiti (no hackfix, but also no CJS configs) ([028c592](https://github.com/flowup/quality-metrics-cli/commit/028c592817b8440e0af5ce1f72e8fffde2f11314))
- configure build and tests to handle ESM and CJS configs ([48cd967](https://github.com/flowup/quality-metrics-cli/commit/48cd967866a84488e6a2382fe44687a31ca47db2))
- **models:** validate plugin icons using portal-client package ([f52cc75](https://github.com/flowup/quality-metrics-cli/commit/f52cc75ed766e47e1d9f82fda1560e0085b3c02e))
- support TS config files using Jiti + hackfix lighthouse import.meta usages ([3b7927d](https://github.com/flowup/quality-metrics-cli/commit/3b7927d65d4607a35dc23d076e72184c281ae8f6))
- **utils:** add file size logging ([#65](https://github.com/flowup/quality-metrics-cli/issues/65)) ([c46046f](https://github.com/flowup/quality-metrics-cli/commit/c46046f9756ea2d02e1d9bb7cc0bbfeff09e61a3)), closes [#59](https://github.com/flowup/quality-metrics-cli/issues/59)
- **utils:** add package `utils` for cli independent logic ([#39](https://github.com/flowup/quality-metrics-cli/issues/39)) ([dacaaf7](https://github.com/flowup/quality-metrics-cli/commit/dacaaf74fb4795a96083ca00fd3b7ca5d3928400)), closes [#32](https://github.com/flowup/quality-metrics-cli/issues/32)
- **utils:** add upload logic ([#66](https://github.com/flowup/quality-metrics-cli/issues/66)) ([3e88ffc](https://github.com/flowup/quality-metrics-cli/commit/3e88ffcbbe375f1fe01bdfd05c088253eeeb98d9)), closes [#21](https://github.com/flowup/quality-metrics-cli/issues/21) [#57](https://github.com/flowup/quality-metrics-cli/issues/57) [#85](https://github.com/flowup/quality-metrics-cli/issues/85)
- **utils:** move parts from utils into core ([#81](https://github.com/flowup/quality-metrics-cli/issues/81)) ([dd0a180](https://github.com/flowup/quality-metrics-cli/commit/dd0a1805ddb97de14d7a4938938aa0bfd852a528)), closes [#78](https://github.com/flowup/quality-metrics-cli/issues/78)

### Bug Fixes

- **cli:** run `npm install` before tests ([6507c2e](https://github.com/flowup/quality-metrics-cli/commit/6507c2e9c2e4105144c03ef74cdfbe1e999355a7))
- lint config and errors ([6f5f677](https://github.com/flowup/quality-metrics-cli/commit/6f5f6779a37359fdde2740fa42e44e7320fa190c))
- refactor after core package intro ([#83](https://github.com/flowup/quality-metrics-cli/issues/83)) ([aa39d09](https://github.com/flowup/quality-metrics-cli/commit/aa39d098c53f7de958509ad465c18b6bee5ec4b9))

### Code Refactoring

- add version targets and ([0d5b87c](https://github.com/flowup/quality-metrics-cli/commit/0d5b87c5a3edbe0fb59dd021dd94779d425ed716))
- **chore:** improvements testing ([#46](https://github.com/flowup/quality-metrics-cli/issues/46)) ([ce21d87](https://github.com/flowup/quality-metrics-cli/commit/ce21d8775e18903adfb4651cbd401d5466592af5))
- **cli:** prepare for core package ([#80](https://github.com/flowup/quality-metrics-cli/issues/80)) ([0c915f0](https://github.com/flowup/quality-metrics-cli/commit/0c915f0dd9b507514e329ff240e0d0511670f2f6)), closes [#73](https://github.com/flowup/quality-metrics-cli/issues/73) [#79](https://github.com/flowup/quality-metrics-cli/issues/79)
- **cli:** rename configPath to config ([#109](https://github.com/flowup/quality-metrics-cli/issues/109)) ([daf548d](https://github.com/flowup/quality-metrics-cli/commit/daf548df1d153409998213ca11914212a1d036b6))
- **models:** add global options ([#76](https://github.com/flowup/quality-metrics-cli/issues/76)) ([28966a2](https://github.com/flowup/quality-metrics-cli/commit/28966a263434ee9e0d4d86466b2d78f389dd6324)), closes [#73](https://github.com/flowup/quality-metrics-cli/issues/73)
- **models:** name file related path more specific ([#102](https://github.com/flowup/quality-metrics-cli/issues/102)) ([fff1ae2](https://github.com/flowup/quality-metrics-cli/commit/fff1ae29c2b3ceb6bc53de1bdaf222859dc9fc83)), closes [#96](https://github.com/flowup/quality-metrics-cli/issues/96)
- **models:** refine data structure ([#55](https://github.com/flowup/quality-metrics-cli/issues/55)) ([f94933b](https://github.com/flowup/quality-metrics-cli/commit/f94933b008fd4b475e6a2ceb1d9d008899df8f53)), closes [#50](https://github.com/flowup/quality-metrics-cli/issues/50)
- move config loading tests to cli-e2e project ([16cd86d](https://github.com/flowup/quality-metrics-cli/commit/16cd86dedf721bf9e0083cc779e754e22b852074))
- purge remaining jest references ([3c6c0ae](https://github.com/flowup/quality-metrics-cli/commit/3c6c0ae74f1374941781088c11abc41d9dee1e7f))
- **utils:** add package.json metadata to report in cli, not utils ([7d3de10](https://github.com/flowup/quality-metrics-cli/commit/7d3de1090cb820121fe245c24d453919546f90e8))

File: nx.json [hunk 1]
```diff
@@ -26,6 +26,26 @@
...
+      "inputs": ["production", "^production"],
+      "executor": "@jscutlery/semver:version",
+      "options": {
+        "preset": {
+          "name": "conventionalcommits",
+          "commitUrlFormat": "{{host}}/{{owner}}/{{repository}}/commit/{{hash}}",
+          "compareUrlFormat": "{{host}}/{{owner}}/{{repository}}/compare/{{previousTag}}...{{currentTag}}",
+          "issueUrlFormat": "{{host}}/{{owner}}/{{repository}}/issues/{{id}}",
+          "types": [
+            { "type": "feat", "section": "Features" },
```
Comment (vmasek) [inline nx.json:41]: what I am also using for dependency updates, nx changes or simillar is `build`, would it make sense to have it here too?

Comment (BioPhoton) [reply nx.json:41]: there is `build`and `ci`

Comment (BioPhoton) [reply nx.json:41]: BTW, Nx released now a new command `nx release` we should test it :)

File: packages/core/project.json [hunk 2]
```diff
@@ -15,6 +15,12 @@
         "esbuildConfig": "esbuild.config.js"
       }
     },
+    "version": { "postTargets": ["core:github"] },
```
Comment (BioPhoton) [inline packages/core/project.json:18]: @edbzn  is there a way to move`"postTargets": ["core:github"]` into `nx.json` target defaults with something like `^github` or so?

**Review by BioPhoton — COMMENTED** [review]
ready for review

File: packages/core/project.json (hunk 2 above)
Comment (edbzn) [reply packages/core/project.json:18]: No, it's not currently supported, but it's something I wanted to implement.

File: nx.json [hunk 3]
```diff
@@ -30,6 +30,35 @@
...
+            { "type": "fix", "section": "Bug Fixes" },
+            { "type": "build", "section": "Dependency updates" },
+            { "type": "docs", "section": "Documentation" },
+            { "type": "refactor", "section": "Code Refactoring" },
+            { "type": "perf", "section": "Performance Improvements" }
+          ]
+        }
+      }
```
Comment (matejchalk) [inline nx.json:55]: Maybe a stupid question, but... do we need to specify all this `preset` on such detail? We already define `commitMessageFormat` explicitly, so what is it used for?

In the [`portal-client`'s `version` executor](https://github.com/code-pushup/portal/blob/main/libs/portal-client/project.json#L64-L68), I only needed to specify `"preset": "conventional"`. That's much less configuration to maintain. :man_shrugging:  (The commit messages are then `chore(portal-client): release version 0.1.2` and the tag is [`portal-client-0.1.2`](https://github.com/code-pushup/portal/releases/tag/portal-client-0.1.2), but I assume only `commitMessageFormat` and `tagPrefix` are needed to customize that).

File: packages/cli/project.json [hunk 4]
```diff
@@ -19,6 +19,14 @@
...
       "dependsOn": ["build"]
     },
+    "version": {
+      "postTargets": ["cli:github"]
+    },
+    "github": {
+      "options": {
+        "notesFile": "./packages/cli/CHANGELOG.md"
+      }
+    },
```
Comment (matejchalk) [inline packages/cli/project.json:29]: What about `ngx-deploy-npm`? Is that another PR? Just checking.

File: packages/models/test/schema.mock.ts
Comment (matejchalk) [inline packages/models/test/schema.mock.ts:1]: Why is this file included?

File: project.json
Comment (matejchalk) [inline project.json:1]: Deleting this is a mistake, I believe. We need it for #33.

File: package.json [hunk 5]
```diff
@@ -85,7 +86,8 @@
     "typescript": "~5.1.3",
     "verdaccio": "^5.0.4",
     "vite": "~4.3.9",
-    "vitest": "~0.32.0"
+    "vitest": "~0.32.0",
+    "webpack-bundle-analyzer": "^4.9.1"
```
Comment (matejchalk) [inline package.json:90]: Why was `webpack-bundle-analyzer` added? We don't use Webpack in this repo.

File: packages/cli/project.json (hunk 4 above)
Comment (BioPhoton) [reply packages/cli/project.json:29]: yes

File: nx.json (hunk 3 above)
Comment (BioPhoton) [reply nx.json:55]: I will check again and come back with a explanation

File: packages/models/test/schema.mock.ts
Comment (BioPhoton) [reply packages/models/test/schema.mock.ts:1]: thanks for spotting!

File: project.json
Comment (BioPhoton) [reply project.json:1]: ok. will then just be explizite with the version target

File: nx.json (hunk 3 above)
Comment (BioPhoton) [reply nx.json:55]: I was able to get rid of some options, but not all...


**By default it is:**

```md
---
## [0.0.2](https://github.com/flowup/quality-metrics-cli/compare/models@0.0.1...models@0.0.2) (2023-11-16)
---
```

**With config it is:**
```md
---
## [0.0.2](https://github.com/flowup/quality-metrics-cli/compare/models@0.0.1...models@0.0.2) (2023-11-16)


### Code Refactoring

* revert ([78d8154](https://github.com/flowup/quality-metrics-cli/commit/78d8154d8494d205b729b2f83ee29c6299b3280e))
---

```

**Comment by BioPhoton** [comment]
What i was wondering is can we reduce the changelog to a specific folder. that would help a lot with duplicates etc in different packages

**Review by matejchalk — APPROVED** [review]

**Comment by matejchalk** [comment]
> What i was wondering is can we reduce the changelog to a specific folder. that would help a lot with duplicates etc in different packages

I don't necessarily mind that there are multiple changelog files :man_shrugging:  It makes sense to distribute it with each NPM package (i.e. it should get copied to `dist`).
